ThirdpartyToolchain.cmake

```diff
@@ -0,0 +1,127 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# MODULE:   ThirdpartyToolchain
+#
+# PROVIDES: resolve_dependency( DEPENDENCY_NAME dependencyName [REQUIRED_VERSION
+# required version] ... )
+#
+# Provides the ability to resolve third party dependencies. If the dependency is
+# already available in the system it will be used.
+#
+# The DEPENDENCY_NAME argument is required. The dependencyName value will be
+# used to search for the installed dependencies Config file and thus this name
+# should match find_package() standards.
+#
+# EXAMPLE USAGE: # Download and setup or use already installed dependency.
+# include(ThirdpartyToolchain) resolve_dependency(folly)
+#
+# ========================================================================================
+
+include(FetchContent)
+
+# =====================================FOLLY==============================================
+
+if(DEFINED ENV{VELOX_FOLLY_URL})
+  set(FOLLY_SOURCE_URL "$ENV{VELOX_FOLLY_URL}")
+else()
+  set(VELOX_FOLLY_BUILD_VERSION v2022.07.11.00)
+  set(FOLLY_SOURCE_URL
+      "https://github.com/facebook/folly/archive/${VELOX_FOLLY_BUILD_VERSION}.tar.gz"
+  )
+  set(VELOX_FOLLY_BUILD_SHA256_CHECKSUM
+      b6cc4082afd1530fdb8d759bc3878c1ea8588f6d5bc9eddf8e1e8abe63f41735)
+endif()
+
+macro(build_folly)
+  message(STATUS "Building Folly from source")
+  # FOLLY_CXX_FLAGS is used internally on folly to define CMAKE_CXX_FLAGS.
+  # Define some known warnings to avoid possible errors on some OS/archs
+  set(FOLLY_CXX_FLAGS -Wno-nullability-completeness -Wno-unused
```

Comment:
Would it be better if we pass these flags in ; Often different architectures require slightly different flags - would be better to centralize this.

Reply:
Maybe these are basic questions:
- Does this mean we need to add architecture specific flags for each dependency ? 
- I suppose this would be set here; I dont see these flags for arm etc (going to test it on the m1 soon) here. 
- Since these flags are distributed around, can there be problems at link time due to incompatible flags ?

Reply:
Not a basic question at all! We don't have to add architecture specific flags for each dependency, the defined `CMAKE_CXX_FLAGS` on `CMakeLists.txt` are the ones used to compile all these third party dependencies which includes the architecture specific ones. We can redefine them here but those ones should be global in my opinion.
As you can see the flags we are defining here are specific known warnings when compiling folly which we want to avoid raising errors as we want to use `-Werror` and `-Wall`. I was just pointing that on Arrow there's some granularity of specific warnings being ignored only for some Architectures.